addons/html_editor/static/src/others/dynamic_placeholder/dynamic_placeholder.js

```diff
@@ -0,0 +1,113 @@
+        validate: Function,
+        resModel: String,
+    };
+}
+
+export class DynamicPlaceholderPlugin extends Plugin {
+    static name = "dynamic_placeholder";
+    static dependencies = ["overlay", "selection", "history", "dom", "qweb"];
+    /** @type { (p: DynamicPlaceholderPlugin) => Record<string, any> } */
+    static resources = (p) => ({
+        powerboxCategory: { id: "marketing_tools", name: _t("Marketing Tools"), sequence: 60 },
+        powerboxCommands: {
+            name: _t("Dynamic Placeholder"),
+            description: _t("Insert personalized content"),
+            category: "marketing_tools",
+            fontawesome: "fa-magic",
+            action(dispatch) {
+                dispatch("OPEN_DYNAMIC_PLACEHOLDER");
+            },
+        },
+    });
+    setup() {
+        this.defaultResModel = this.config.dynamicPlaceholderResModel;
+
+        /** @type {import("@html_editor/core/overlay_plugin").Overlay} */
+        this.overlay = this.shared.createOverlay(EditorDynamicPlaceholder, {
+            position: "bottom-start",
+        });
+        this.overlayProps = {
+            close: this.onClose.bind(this),
+            validate: this.onValidate.bind(this),
+        };
+
+        this.addDomListener(this.document, "click", () => {
+            if (this.overlay.isOpen) {
+                this.overlay.close();
+            }
+        });
+
+        this.addDomListener(
+            this.document,
+            "beforeinput",
+            (e) => {
+                if (this.overlay.isOpen && e.inputType === "insertParagraph") {
+                    e.preventDefault();
+                    e.stopImmediatePropagation();
+                }
+            },
+            true
+        );
+    }
+
+    handleCommand(command, payload) {
+        switch (command) {
+            case "OPEN_DYNAMIC_PLACEHOLDER": {
+                this.open(payload?.resModel || this.defaultResModel);
+                break;
+            }
+        }
+    }
+
+    /**
+     * @param {string} resModel
+     */
+    open(resModel) {
+        this.preservedSelection = this.shared.preserveSelection();
+        this.overlay.open({
+            props: {
+                ...this.overlayProps,
+                resModel: resModel,
+            },
+        });
+    }
+
+    /**
+     * @param {string} chain
+     * @param {string} defaultValue
+     */
+    onValidate(chain, defaultValue) {
+        if (!chain) {
+            return;
+        }
+
+        let dynamicPlaceholder = "object." + chain;
+        dynamicPlaceholder +=
+            defaultValue && defaultValue !== "" ? ` or '''${defaultValue}'''` : "";
+        const t = document.createElement("T");
+        t.setAttribute("t-out", dynamicPlaceholder);
+
+        this.shared.domInsert(t);
+        this.editable.focus();
+        this.dispatch("NORMALIZE", { node: t.parentElement });
+        this.dispatch("ADD_STEP");
+    }
+    onClose() {
+        this.preservedSelection?.restore();
```

Comment:
How it s possible to have undefined preservedSelection ?

Reply:
it shouldn't. 
I just prefer to ensure we avoid trace-back if for some reason you mange to have to system call the close method twice. maybe if you click outside at the same time than you validate the dynamic placeholder.
I can remove it if it bother you.

Reply:
I prefer to remove it. If we have a trace back, we probably have a bug and we should fix it correctly with a test :)